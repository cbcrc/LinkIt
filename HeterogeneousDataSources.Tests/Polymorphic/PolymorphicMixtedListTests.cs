using System.Collections.Generic;
using ApprovalTests.Reporters;
using HeterogeneousDataSources.ConfigBuilders;
using HeterogeneousDataSources.LinkedSources;
using HeterogeneousDataSources.Protocols;
using HeterogeneousDataSources.Tests.Shared;
using NUnit.Framework;
using RC.Testing;

namespace HeterogeneousDataSources.Tests.Polymorphic {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class PolymorphicMixtedListTests {
        private FakeReferenceLoader<Model, string> _fakeReferenceLoader;
        private LoadLinkProtocol _sut;

        [SetUp]
        public void SetUp() {
            var loadLinkProtocolBuilder = new LoadLinkProtocolBuilder();

            loadLinkProtocolBuilder.For<LinkedSource>()
                .PolymorphicLoadLinkForList(
                    linkedSource => linkedSource.Model.TargetReference,
                    linkedSource => linkedSource.Target,
                    link => link.GetType(),
                    includes => includes
                        .Include<Person>().AsReference(
                            typeof(int),
                            link=>link.ToString()
                        )
                        .Include<PersonLinkedSource>().AsNestedLinkedSource(
                            typeof(string),
                            link => link.ToString()
                        )
                        .Include<PersonLinkedSource>().AsSubLinkedSource(
                            typeof(Person),
                            link => (Person)link
                        )
                );
            loadLinkProtocolBuilder.For<PersonLinkedSource>()
                .LoadLinkReferenceById(
                    linkedSource => linkedSource.Model.SummaryImageId,
                    linkedSource => linkedSource.SummaryImage
                );

            _fakeReferenceLoader =
                new FakeReferenceLoader<Model, string>(reference => reference.Id);
            _sut = loadLinkProtocolBuilder.Build(_fakeReferenceLoader);
        }

        [Test]
        public void LoadLink_PolymorphicMixteList() {
            _fakeReferenceLoader.FixValue(
                new Model {
                    Id = "1",
                    TargetReference = new List<object>{
                        1,
                        "nested",
                        new Person{
                            Id = "as-sub-linked-source",
                            Name = "The Name",
                            SummaryImageId = "the-id"
                        }
                    }
                }
            );

            var actual = _sut.LoadLink<LinkedSource>().ById("1");

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        public class LinkedSource : ILinkedSource<Model> {
            public Model Model { get; set; }
            public List<object> Target { get; set; }
        }

        public class Model{
            public string Id { get; set; }
            public List<object> TargetReference { get; set; }
        }
    }
}
