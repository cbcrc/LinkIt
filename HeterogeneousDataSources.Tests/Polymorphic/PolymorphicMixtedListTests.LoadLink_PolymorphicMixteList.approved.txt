{
  "model": {
    "id": "1",
    "targetReference": [
      1,
      "nested",
      {
        "$type": "HeterogeneousDataSources.Tests.Shared.Person, HeterogeneousDataSources.Tests",
        "id": "as-sub-linked-source",
        "name": "The Name",
        "summaryImageId": "the-id"
      }
    ]
  },
  "target": [
    {
      "$type": "HeterogeneousDataSources.Tests.Shared.Person, HeterogeneousDataSources.Tests",
      "id": "1",
      "name": "name-1",
      "summaryImageId": "person-img-1"
    },
    {
      "$type": "HeterogeneousDataSources.Tests.PersonLinkedSource, HeterogeneousDataSources.Tests",
      "model": {
        "id": "nested",
        "name": "name-nested",
        "summaryImageId": "person-img-nested"
      },
      "summaryImage": {
        "id": "person-img-nested",
        "alt": "alt-person-img-nested"
      }
    },
    {
      "$type": "HeterogeneousDataSources.Tests.PersonLinkedSource, HeterogeneousDataSources.Tests",
      "model": {
        "id": "as-sub-linked-source",
        "name": "The Name",
        "summaryImageId": "the-id"
      },
      "summaryImage": {
        "id": "the-id",
        "alt": "alt-the-id"
      }
    }
  ]
}